pkg/migration-tool/reader/reader.go

```diff
@@ -93,5 +117,133 @@ func (rr *RemoteRead) Run(errChan chan<- error) {
 			rr.sigBlockRead <- blockRef
 			blockRef = nil
 		}
-	}()
+	}
+
+	readPromBlocks := func() {
+		// Comment.
+		defer deferFunc()
+		db, err := tsdb.OpenDBReadOnly(rr.path, nil)
+		if err != nil {
+			errChan <- fmt.Errorf("remote-read run: opening prom tsdb: %w", err)
+			return
+		}
+		defer func() {
+			err = tsdb_errors.NewMulti(err, db.Close()).Err()
+		}()
+		querier, err := db.Querier(rr.c, rr.plan.Mint(), rr.plan.Maxt())
+		if err != nil {
+			errChan <- fmt.Errorf("remote-read run: prom tsdb querier: %w", err)
+			return
+		}
+		defer querier.Close()
+		ss := querier.Select(false, nil, labels.MustNewMatcher(labels.MatchRegexp, "", ".*"))
+		var iteratorStore = new(sync.Map)
+		type cusMap struct {
+			labels labels.Labels
+			itr    chunkenc.Iterator
+		}
+		// Fill the iteratorStore with the iterators across all the series-sets for the given time-range.
+		for ss.Next() {
+			series := ss.At()
+			iteratorStore.Store(series.Labels().String(), cusMap{labels: series.Labels(), itr: series.Iterator()})
+		}
+		// Start creating blocks and filling them with samples.
+		for rr.plan.ShouldProceed() {
+			select {
+			case <-rr.c.Done():
+				return
+			case <-rr.SigForceStop:
+				return
+			default:
+			}
+			blockRef, err := rr.plan.NextBlock()
+			if err != nil {
+				errChan <- fmt.Errorf("remote-run run: %w", err)
+				return
+			}
+			ts := make(tsMap)
+			mint, maxt := blockRef.Mint(), blockRef.Maxt()
+			callBack := func(_ interface{}, value interface{}) bool {
+				cus, ok := value.(cusMap)
+				if !ok {
+					errChan <- fmt.Errorf("remote-read run: value not found in iteratorStore")
+					return false
+				}
+				itr := cus.itr
+				var samples []prompb.Sample
+				appender := func() bool {
+					t, v := itr.At()
+					if t >= mint && t < maxt {
```

Comment:
shouldn't this be guaranteed by the iterator?

Reply:
I don't think so. The iterator will be per series basis, but the samples iteration isn't within the time-range. It happens by `Seek()` which goes to the time required and iterates till the last `Next()` (stopped by break here) and stays there as cache in the iteratorStore (so that for the next time-range delta, it can resume from there)